### PR TITLE
How to use PostgreSQL provider on Aiven PG DB. For issue-1151

### PR DESCRIPTION
**Pull request by jorgeuribe-aiven**
# What changed, and why it matters

I have added steps for performing 3rd party provider configuration, documentation is not clear enough and some people find hard to work out how to configure it.  

Fixes https://github.com/aiven/devportal/issues/1151

### CODE REVIEW

**Comment by ftisiot** [comment]
Hi @jorgeuribe-aiven, thanks for the contribution! I believe we already have something similar in our recipes for PostgreSQL, would it make sense to add also this article as a recipe? Tagging @dewan-ahmed since he was managing the cookbook

**Comment by dewan-ahmed** [comment]
I believe that the intent is to show other Terraform provider (something generic to PG) to create an Aiven resource. I don't think this qualifies for a recipe since the [cookbook](https://developer.aiven.io/docs/tools/terraform/reference/cookbook.html) focuses on Aiven Terraform Provider only. 

My question is: do we want to have non-Aiven Terraform provider on our developer documentation? Is there something specific that this generic provider can do that Aiven Terraform Provider cannot?

**Comment by jorgeuribe-aiven** [comment]
If I am right, Aiven provider only support PG service, database and user resources, whereas the postgresql provider offers these resources: 

postgresql_database
postgresql_default_privileges
postgresql_extension
postgresql_function
postgresql_grant
postgresql_grant_role
postgresql_physical_replication_slot
postgresql_publication
postgresql_replication_slot
postgresql_role
postgresql_schema

I came across the postgresql provider when customer had some problems configuring it for an Aiven PG service.

**Comment by lornajane** [comment]
I think that this looks valuable to include if there are things that people can't do with our provider but could achieve by using a combination of them. Agree it's not a cookbook but could we please review and make clear when the additional provider is useful and what problem it solves?

File: _toc.yml [hunk 1]
```diff
@@ -132,6 +132,8 @@ entries:
           entries:
             - file: docs/tools/terraform/howto/upgrade-provider-v1-v2
               title: Upgrade the Aiven Terraform Provider from v1 to v2
+            - file: docs/tools/terraform/howto/config-postgresql-provider
+              title: Configure PostgreSQL provider for Aiven PG services
```
Comment (dewan-ahmed) [inline _toc.yml:136]: Proposed title: Use PostgreSQL provider alongside Aiven Terraform Provider

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 2]
```diff
@@ -0,0 +1,80 @@
+Use PostgreSQL® provider with Aiven's PostgreSQL service
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:1]: Title suggestion: Use PostgreSQL® provider for additional PostgreSQL configurations

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#####################################################
+
+Aiven's Terraform provider is aimed for provisioning and performing basic service configuration, if additional configuration on the service is required there are other providers that can be used to perform the task.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:4]: Suggested text:

[Aiven Terraform Provider](https://registry.terraform.io/providers/aiven/aiven/latest/docs) can be used to create and manage Aiven for PostgreSQL® service, PostgreSQL® databases, and users. If you need to perform additional configurations such as setting PostgreSQL default privileges, configure PostgreSQL publication, or reuse a PostgreSQL-based sub-module between different vendors to make the Terraform code homogeneous., you can consider using the [PostgreSQL® provider](https://registry.terraform.io/providers/cyrilgdn/postgresql/latest/docs).

Comment (dewan-ahmed) [reply docs/tools/terraform/howto/config-postgresql-provider.rst:4]: I used the markdown format for the links. Please remember to change to RST format.

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 4]
```diff
@@ -0,0 +1,80 @@
+Use PostgreSQL® provider with Aiven's PostgreSQL service
+#####################################################
+
+Aiven's Terraform provider is aimed for provisioning and performing basic service configuration, if additional configuration on the service is required there are other providers that can be used to perform the task. 
+This article shows how to use the `PostgreSQL® provider <https://registry.terraform.io/providers/cyrilgdn/postgresql/latest/docs>`_ along with the ``Aiven Postgres service``.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:5]: Suggested text: This article shows how to use the PostgreSQL provider along with the Aiven Terraform Provider to create a PostgreSQL role.

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+#####################################################
+
+Aiven's Terraform provider is aimed for provisioning and performing basic service configuration, if additional configuration on the service is required there are other providers that can be used to perform the task. 
+This article shows how to use the `PostgreSQL® provider <https://registry.terraform.io/providers/cyrilgdn/postgresql/latest/docs>`_ along with the ``Aiven Postgres service``.   
+
+
+Configure the required providers
+--------------------------------
+
+The new provider must be added to the ``required providers`` block in the Terraform code.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:11]: ``required_providers``

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+1. This example shows how to add the ``PostgreSQL`` provider along with the ``Aiven`` provider
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:13]: This example shows how to add the PostgreSQL provider (source: ``cyrilgdn/postgresql``) along with the ``Aiven Terraform Provider`` (source: ``aiven/aiven``).

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          version = ">= 3.1"
+        }
+        postgresql = {
+          source  = "cyrilgdn/postgresql"
+          version = "1.16.0"
+        }        
+      }
+    }
+
+2. Then configure the provider with the corresponding information from the ``Aiven Postgres service``.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:30]: If the PostgreSQL provider is used on its own, you can provide the Aiven for PostgreSQL service connection details as follows:

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      host            = "pg-serivicename-projectname.aivencloud.com"
+      port            = 12691
+      database        = "defaultdb"
+      username        = "avnadmin"
+      password        = "postgres_password"
+      sslmode         = "require"
+      connect_timeout = 15
+    }
+
+Optionally, when the Aiven PG service is created in the same Terraform project the values required to configure the ``PostgreSQL`` provider can be passed using references to the resource, as shown in the code below:
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:44]: Optionally, when the Aiven for PostgreSQL service is created within the same Terraform project, the values required to configure the PostgreSQL provider can be passed using references to the resource, as shown in the code below:

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+.. code:: terraform
+
+    # A new Aiven PG service is created.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:48]: The Terraform code is self-explanatory so a comment is not required. If you still keep the comment, please rewrite as:

# Aiven for PostgreSQL service on GCP

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    resource "aiven_pg" "pg-service" {
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:49]: resource "aiven_pg" "demo-pg" {
  project                = var.project_name
  cloud_name             = "google-asia-southeast1"
  plan                   = "business-8"
  service_name           = "demo-pg"
  termination_protection = true
}

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      project                 = "my_project"
+      cloud_name              = "cloud_name"
+      plan                    = "plan"
+      service_name            = "pg-myservice"
+    }
+
+    # References to the aiven_pg.pg-service are used for configuring the provider.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:56]: `# PostgreSQL provider is configured with references to the aiven_pg.demo-pg resource.`

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    provider "postgresql" {
+      host            = aiven_pg.pg-service.service_host
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:58]: host            = aiven_pg.demo-pg.service_host

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      port            = aiven_pg.pg-service.service_port
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:59]: port            = aiven_pg.demo-pg.service_port

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      database        = "defaultdb"
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:60]: database        = aiven_pg.demo-pg.pg.dbname

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      username        = aiven_pg.pg-service.service_username
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:61]: username        = aiven_pg.demo-pg.service_username

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+      password        = aiven_pg.pg-service.service_password
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:62]: password        = aiven_pg.demo-pg.service_password

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+      sslmode         = "require"
+      connect_timeout = 15
+    }
+
+3. Start using the resources available for the ``PostgreSQL`` provider. The following example shows how to create a role.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:67]: 3. Create a PostgreSQL role called **test_role** using the Terraform resource `postgresql_role.my_role`.

File: docs/tools/terraform/howto/config-postgresql-provider.rst [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+.. code:: terraform
+
+    resource "postgresql_role" "my_role" {
+      name     = "test_role"
+    }
+
+.. note::
+
+  For the full documentation of the ``Aiven provider`` refer to `Aiven provider documentation <https://registry.terraform.io/providers/aiven/aiven/latest/docs>`_.
```
Comment (dewan-ahmed) [inline docs/tools/terraform/howto/config-postgresql-provider.rst:77]: Full documentation of the `Aiven Terraform Provider <https://registry.terraform.io/providers/aiven/aiven/latest/docs>`_.

**Review by dewan-ahmed — COMMENTED** [review]
Hi @jorgeuribe-aiven I added some review comments for revising some texts (for example, using the correct product name, adding ®, etc.) and updating the service name to keep consistency (for example, we use ``demo-XXX`` for service names in other Terraform-related articles). 

Looking forward to this addition to our developer documentation. Thanks so much for your contributions.

File: _toc.yml (hunk 1 above)
Comment (jorgeuribe-aiven) [reply _toc.yml]: Hi @dewan-ahmed I have make the changes you have suggested and pushed them. I will resolve all the conversations.

**Review by dewan-ahmed — APPROVED** [review]
Thanks for this excellent addition to our Terraform HowTo @jorgeuribe-aiven 🎉

**Pull request merged**